### ISSUE TITLE
Fix/update IRC reply constants

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8585]
|[<img alt="scop's avatar" src="https://avatars.githubusercontent.com/u/109152?s=50" width="50" height="50">](https://github.com/scop)| @scop reported|
|-|-|
|Trac ID|trac#8585|
|Type|defect|
|Created|2016-07-10 09:15:20Z|
|Branch|https://github.com/twisted/twisted/tree/8585-scop-ircconstants|

The IRC reply constants erroneously list RPL_ADMINLOC twice with two different numbers. There is no such reply as RPL_ADMINLOC in the first place, RPL_ADMINLOC1 and RPL_ADMINLOC2 on the other hand do exist (but are missing from twisted). RPL_ADMINLOC should be (eventually, maybe with deprecation) removed, and RPL_ADMINLOC1 (number 257) and RPL_ADMINLOC2 (number 258) added.

https://tools.ietf.org/html/rfc2812

ERR_TOOMANYMATCHES (number 416) from RFC 2812 errata is missing, it should be added.

https://www.rfc-editor.org/errata_search.php?rfc=2812&eid=2822

<details><summary>Searchable metadata</summary>

```
trac-id__8585 8585
type__defect defect
reporter__scop scop
priority__low low
milestone__None None
branch__8585_scop_ircconstants 8585-scop-ircconstants
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__None None
time__1468142120936769 1468142120936769
changetime__1474198569965514 1474198569965514
version__None None
owner__GitHub__noreply_____ GitHub <noreply@...>
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="scop's avatar" src="https://avatars.githubusercontent.com/u/109152?s=50" width="50" height="50">](https://github.com/scop)<a name="note_2"></a>|@scop commented|
|-|-|

https://github.com/twisted/twisted/pull/314

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @scop|
|-|-|

Hi,

Thanks for the changes.

I don't understand the scope of this branch.

It has a .misc, but to me it looks like a .bugfix and a .feature

Please note that I have little information about how IRC works and how words IRC is designed.

Why do we need to add ERR_TOOMANYMATCHES if it is not used anywhere else in the code?

Also RPL_ADMINLOC is a public API so we should deprecate it first before removing it... but why do we need to remove it. Why do we need 2 values ?

Regarding the deprecation part, please see our documentation and the motivation behind the whole deprecation dance http://twistedmatrix.com/documents/current/core/development/policy/compatibility-policy.html

Please add more info about what is broken, why is broken and what is the scope of this branch.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="scop's avatar" src="https://avatars.githubusercontent.com/u/109152?s=50" width="50" height="50">](https://github.com/scop)<a name="note_4"></a>|@scop removed owner|
|-|-|

Replying to [adiroiban](#note_3):
> Why do we need to add ERR_TOOMANYMATCHES if it is not used anywhere else in the code?

1) For consistency, because all the other IRC ERR_* and RPL_* constants are there, even if not used in the code.

2) AFAIU, when properly mapped, this will allow to implement an irc_ERR_TOOMANYMATCHES() method for handling it.

> Also RPL_ADMINLOC is a public API so we should deprecate it first before removing it... but why do we need to remove it. Why do we need 2 values ?

Because that's the way they're specified in the RFC. I've updated the bug description, hopefully that clarifies it.

> Please add more info about what is broken, why is broken and what is the scope of this branch.

Hopefully the updated descriptions suffice. The lowest hanging broken bit is that in the current code, the (incorrect in the first place) RPL_ADMINLOC is being defined twice with two different values and added to a dict twice with two different values. Obviously, only the latter ends up being in effect, and thus number 257 is left without a mapping altogether.

Anyway, I've now spent the time I want to spend on this, and going through the deprecation dance is too much work for a thing that I don't actually use, nor really care about, just happened to notice. So I'm canceling the pull request and leaving this as a regular bug report, with instructions how it should be fixed now in the description. I'll delete the PR soonish, but before that will let it hang out there for a bit if someone wants to grab the code.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Thanks for the update.

It make sense now.

So, if there is no such thing as RPL_ADMINLOC then I think that this is a gross violation of the specification and we can change it without the deprecation dance.

So this branch is both a bugfix and a feature.

Since the changes are minor I think that we can merge them in a single branch.

But since this is a bugfix and a feature it needs a release notes fragment.

I will prepare the branch and send it to buildbot.

----

I know that it is a lot of work for such a small change, but I think that the release notes are important and useful and we should not skip them.

-----

Since I have also changed this branch, it needs someone else to check that the release notes fragments make sense.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|GitHub set owner to GitHub<br>GitHub set status to `closed`|
|-|-|

In changeset bd18732299cbf4438d774a4a866b6b286891304f
```
#!CommitTicketReference repository="" revision="bd18732299cbf4438d774a4a866b6b286891304f"
Merge pull request #355 from twisted/8585-scop-ircconstants

Author: scop
Reviewer: adiroiban, rodrigc
Fixes: #8585

Add IRC constants specified in RFC 2812 and errata.
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing the review tag from closed tickets
